src/main/scala/com.snowplowanalytics/iglu.schemaddl/migrations/SchemaList.scala

```diff
@@ -0,0 +1,286 @@
+  * Schemas in the list will be in correct order.
+  */
+sealed trait SchemaList extends Product with Serializable
+
+object SchemaList {
+
+  /**
+    * Always belong to the same vendor/name/model triple,
+    * always have at least two elements: head (initial x-0-0 schema) and latest schema
+    */
+  case class SchemaListFull private(schemas: NonEmptyList[IgluSchema]) extends SchemaList {
+
+    /**
+      * Create SchemaListSegment from schemas in current SchemaListFull
+      */
+    def toSegment: SchemaListSegment = SchemaListSegment(schemas)
+
+    /**
+      * Create segments from all possible combinations of NonEmptyList(source,between,destination),
+      */
+    def extractSegments: NonEmptyList[SchemaListSegment] = {
+      val res = SchemaList.buildMatrix(schemas.toList).map { case(_, _, nel) => SchemaListSegment(nel) }
+      NonEmptyList.fromListUnsafe(res)
+    }
+
+    /**
+      * Create new segment with items after index
+      */
+    def afterIndex(i: Int): Option[SchemaListSegment] =
+      schemas.zipWithIndex.filter { case (_, c) => c >= i }.map(_._1) match {
+        case h :: t => Some(SchemaListSegment(NonEmptyList(h, t)))
+        case _ => None
+      }
+  }
+  private[migrations] object SchemaListFull
+
+  /**
+    * Represents schema list with single item.
+    * It's version is always first of the model group
+    * such that 1-0-0, 2-0-0 etc.
+    */
+  case class SingleSchema private(schema: IgluSchema) extends SchemaList
+  private[migrations] object SingleSchema
+
+  /**
+    * Has all properties of SchemaListFull except that it can miss initial or last schemas
+    */
+  case class SchemaListSegment private(schemas: NonEmptyList[IgluSchema]) extends AnyVal
+  private[migrations] object SchemaListSegment
+
+  /**
+    * Represents errors while creating SchemaFullList
+    */
+  sealed trait FullListCreationError extends Product with Serializable
+
+  object FullListCreationError {
+
+    /**
+      * Returned when given model group have schemas which
+      * they could not be ordered unambiguously.
+      * For example, [1-0-0, 1-1-0, 1-0-2] schema list
+      * could not be ordered unambiguously because it could be
+      * either [1-0-0, 1-0-2, 1-1-0] or [1-0-0, 1-1-0, 1-0-2]
+      */
+    case class AmbiguousOrder(schemas: ModelGroupList) extends FullListCreationError
+
+    /**
+      * Returned when there is a gap in the schema list.
+      * [1-0-0, 1-0-1, 1-0-3] is missing 1-0-2 version,
+      * therefore gap error will be returned in this case
+      */
+    case class GapInModelGroup(schemas: ModelGroupList) extends FullListCreationError
+
+    /**
+      * Represents unexpected error cases while creating SchemaFullList
+      */
+    case class UnexpectedError(schemas: ModelGroupList) extends FullListCreationError
+  }
+
+  /**
+    * Always belong to the same vendor/name/model triple,
+    * however different from SchemaList, no ordering is implied.
+    */
+  sealed trait ModelGroupList {
```

Comment:
Why it isn't a case class? We'd get `equals` for free and it would be much shorter.

Reply:
I've used trait in here in order to use smart constructor pattern which mentioned in [this book](https://www.manning.com/books/functional-and-reactive-domain-modeling).

Reply:
But we can use smart constructors without being restricted to sealed traits. The book is very good, but community found out a lot of better patterns since then. I usually go with private constructors and in most cases that's fine. The most bullet-proof approach I know is this one: https://gist.github.com/tpolecat/a5cb0dc9adeacc93f846835ed21c92d2.

Sealed traits add safety, but significantly increase boilerplate (wondering what `toString` looks like). I don't think this is a good trade-off.

Reply:
Okay then, I will take a look at the gist you provided 👍

Reply:
Feel free to switch all my private-constructor classes into this approach. I just been too lazy to apply it.